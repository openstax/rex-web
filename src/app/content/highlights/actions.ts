import { Highlight, NewHighlight, UpdateHighlightRequest } from '@openstax/highlighter/dist/api';
import { createStandardAction } from 'typesafe-actions';
import { LocationFilters } from '../components/popUp/types';
import {
  CountsPerSource,
  HighlightData,
  SummaryFilters,
  SummaryFiltersUpdate,
  SummaryHighlights,
  SummaryHighlightsPagination,
} from './types';

export const focusHighlight = createStandardAction('Content/Highlights/focus')<string>();
export const clearFocusedHighlight = createStandardAction('Content/Highlights/clear')();
export const createHighlight = createStandardAction('Content/Highlights/create')<NewHighlight, {
  revertingAfterFailure?: boolean,
  locationFilterId: string,
  pageId: string,
}>();
export const requestDeleteHighlight = createStandardAction('Content/Highlights/requestDelete')<Highlight, {
  locationFilterId: string,
  pageId: string,
}>();
export const receiveDeleteHighlight = createStandardAction('Content/Highlights/receiveDelete')<Highlight, {
  revertingAfterFailure?: boolean,
  locationFilterId: string,
  pageId: string,
}>();
export const updateHighlight = createStandardAction('Content/Highlights/update')<UpdateHighlightRequest, {
  revertingAfterFailure?: boolean,
  preUpdateData: UpdateHighlightRequest,
  locationFilterId: string,
  pageId: string,
}>();
export const receiveHighlights = createStandardAction(
  'Content/Highlights/receive'
)<{highlights: HighlightData[], pageId: string}>();
export const setAnnotationChangesPending = createStandardAction('Content/Highlights/setAnnotationChangesPending')<
  boolean
>();

export const openMyHighlights = createStandardAction('Content/Highlights/Summary/open')<void>();
export const closeMyHighlights = createStandardAction('Content/Highlights/Summary/close')<void>();
export const initializeMyHighlightsSummary = createStandardAction('Content/Highlights/Summary/init')<void>();

export const printSummaryHighlights = createStandardAction('Content/Highlights/Summary/print')();
export const toggleSummaryHighlightsLoading = createStandardAction('Content/Highlights/Summary/loading')<boolean>();

export const loadMoreSummaryHighlights = createStandardAction('Content/Highlights/Summary/loadMore')();
export const setSummaryFilters = createStandardAction('Content/Highlights/Summary/setFilters')<
  Partial<SummaryFilters>
>();
export const updateSummaryFilters = createStandardAction('Content/Highlights/Summary/updateFilters')<
  Partial<SummaryFiltersUpdate>
>();
export const receiveSummaryHighlights = createStandardAction('Content/Highlights/Summary/receiveHighlights')<
  SummaryHighlights,
  {
    pagination: SummaryHighlightsPagination,
    filters?: SummaryFilters,
    isStillLoading?: boolean
  }
>();
export const receiveHighlightsTotalCounts = createStandardAction(
  'Content/receiveHighlightsTotalCounts'
)<CountsPerSource, LocationFilters>();
