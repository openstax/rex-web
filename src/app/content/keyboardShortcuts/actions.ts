import { createStandardAction } from 'typesafe-actions';

export const openKeyboardShortcutsMenu = createStandardAction('Content/KeyboardShortcuts/Menu/open')<void>();
export const closeKeyboardShortcutsMenu = createStandardAction('Content/KeyboardShortcuts/Menu/close')<void>();
