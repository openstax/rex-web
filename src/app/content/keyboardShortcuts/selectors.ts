import { createSelector } from 'reselect';
import * as navigationSelectors from '../../navigation/selectors';
import { modalUrlName } from './constants';

export const isKeyboardShortcutsOpen = createSelector(
  navigationSelectors.query, (query) => !!query && query.modal === modalUrlName
);
