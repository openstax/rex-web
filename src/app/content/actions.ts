import { createStandardAction } from 'typesafe-actions';
import * as highlightingActions from './highlights/actions';
import * as practiceQuestionsActions from './practiceQuestions/actions';
import * as searchActions from './search/actions';
import * as studyGuidesActions from './studyGuides/actions';
import { ArchivePage, Book, Params, State } from './types';

export const search = searchActions;
export const highlighting = highlightingActions;
export const studyGuides = studyGuidesActions;
export const practiceQuestions = practiceQuestionsActions;

export const openToc = createStandardAction('Content/openToc')<void>();
export const closeToc = createStandardAction('Content/closeToc')<void>();
export const resetToc = createStandardAction('Content/resetToc')<void>();

export const receiveBuyPrintConfig = createStandardAction('Content/receiveBuyPrintConfig')<State['buyPrint']>();

export const requestBook = createStandardAction('Content/requestBook')<Params['book']>();
export const receiveBook = createStandardAction('Content/receiveBook')<Book>();

export const requestPage = createStandardAction('Content/requestPage')<Params['page']>();
export const receivePage = createStandardAction('Content/receivePage')<
  ArchivePage & {references: State['references']}
>();
export const receivePageNotFoundId = createStandardAction('Content/receivePageNotFoundId')<string>();

export const openNudgeStudyTools = createStandardAction('Content/openNudgeStudyTools')();
export const closeNudgeStudyTools = createStandardAction('Content/closeNudgeStudyTools')();
