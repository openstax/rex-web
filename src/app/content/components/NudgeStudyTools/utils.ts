import { HTMLElement } from '@openstax/types/lib.dom';
import * as Cookies from 'js-cookie';
import React, { useState } from 'react';
import { useSelector } from 'react-redux';
import { useDebouncedWindowSize, useOnScrollTopOffset } from '../../../reactUtils';
import { assertDocument, remsToPx } from '../../../utils';
import { page as pageSelector } from '../../selectors';
import { hasStudyGuides } from '../../studyGuides/selectors';
import { Page } from '../../types';
import {
  arrowDesktopHeight,
  arrowRightMargin,
  arrowTopMargin,
  closeButtonDistanceFromContent,
  closeButtonSize,
  contentMarginLeft,
  contentMarginTop,
  contentWidth,
  cookieNudge,
  daysUntilCookieExpires,
  mobileNudgeStudyToolsTargetId,
  mobileSpotlightPadding,
  nudgeStudyToolsMinPageLimit,
  nudgeStudyToolsShowLimit,
  nudgeStudyToolsTargetId,
  timeIntervalBetweenShowingNudgeInMs,
} from './constants';

interface Positions {
  arrowLeft: number;
  arrowTopOffset: number;
  closeButtonLeft: number;
  closeButtonTopOffset: number;
  contentWrapperLeft: number;
  contentWrapperTopOffset: number;
  spotlightHeight: number;
  spotlightLeftOffset: number;
  spotlightTopOffset: number;
  spotlightWidth: number;
}

export const getPositions = (target: HTMLElement, isMobile: boolean): Positions => {
  const padding = remsToPx(mobileSpotlightPadding);
  const { top, left, height, width } = getBoundingRectOfNudgeTarget(target);
  const spotlightLeftOffset = left - (isMobile ? padding : 0);
  const spotlightWidth = width + (isMobile ? padding * 2 : 0);
  const arrowTopOffset = top + height + remsToPx(arrowTopMargin);
  // left edge of arrow image should be on the middle of the spotlight (adjusted for a margin)
  const arrowLeft = spotlightLeftOffset + (spotlightWidth / 2) - remsToPx(arrowRightMargin);

  const contentWrapperTopOffset = arrowTopOffset
    + remsToPx(arrowDesktopHeight)
    + remsToPx(contentMarginTop);
  const contentWrapperLeft = spotlightLeftOffset + spotlightWidth + remsToPx(contentMarginLeft);
  const closeButtonLeft = contentWrapperLeft + remsToPx(contentWidth) + remsToPx(closeButtonDistanceFromContent);
  const closeButtonTopOffset = contentWrapperTopOffset
    - remsToPx(closeButtonSize)
    - remsToPx(closeButtonDistanceFromContent);

  return {
    arrowLeft,
    arrowTopOffset,
    closeButtonLeft,
    closeButtonTopOffset,
    contentWrapperLeft,
    contentWrapperTopOffset,
    spotlightHeight: height,
    spotlightLeftOffset,
    spotlightTopOffset: top,
    spotlightWidth,
  };
};

// Target may have display set to `contents` so we are calculatings rect for children of passed target.
const getBoundingRectOfNudgeTarget = (target: HTMLElement) => {
  const { top, bottom, left, right } = Array.from(target.children).reduce((rects, child) => {
    const rect = child.getBoundingClientRect();
    rects.top = Math.min(rects.top, rect.top);
    rects.bottom = Math.max(rects.bottom, rect.bottom);
    rects.left = Math.min(rects.left, rect.left);
    rects.right = Math.max(rects.right, rect.right);
    return rects;
    // Starting values depends on if we checks for Math.max or Math.min
  }, { top: 9999, bottom: 0, left: 9999, right: 0 });

  return {
    height: bottom - top,
    left,
    right,
    top,
    width: right - left,
  };
};

export const usePositions = (isMobile: boolean) => {
  const [windowWidth] = useDebouncedWindowSize();
  const scrollTopOffset = useOnScrollTopOffset();
  const [positions, setPositions] = React.useState<Positions | null>(null);
  const studyGuides = useSelector(hasStudyGuides);

  React.useEffect(() => {
    const document = assertDocument();
    const target = document.querySelector(
      `#${isMobile ? mobileNudgeStudyToolsTargetId : nudgeStudyToolsTargetId}`
    ) as HTMLElement | null;
    if (target) {
      // Make sure that we calculate positions with body overflow set to hidden
      // because it causes scrollbar to hide which results in different positions.
      const prevOverflow = document.body.style.overflow;
      document.body.style.overflow = 'hidden';
      setPositions(getPositions(target, isMobile));

      // Resets to the value from before calculations. We want this style change to be handled
      // directly in the component.
      document.body.style.overflow = prevOverflow;
    }
    return () => setPositions(null);
  }, [scrollTopOffset, studyGuides, windowWidth, isMobile]);

  return positions;
};

export const getCounterCookie = () => {
  return Number(Cookies.get(cookieNudge.counter) || 0);
};

export const getDateCookie = () => {
  const lastShownDate = Cookies.get(cookieNudge.date);
  return lastShownDate ? new Date(lastShownDate) : undefined;
};

export const passedTimeInterval = () => {
  const lastShownDate = getDateCookie();
  return !lastShownDate
    ? true
    : (Date.now() - lastShownDate.getTime()) > timeIntervalBetweenShowingNudgeInMs
  ;
};

export const getPageCounterCookie = () => {
  return Number(Cookies.get(cookieNudge.pageCounter) || 0);
};

export const incrementPageCounterCookie = () => {
  const counter = getPageCounterCookie();
  const newValue = counter + 1;
  Cookies.set(cookieNudge.pageCounter, newValue.toString(), {expires: daysUntilCookieExpires});
  return newValue;
};

export const shouldDisplayNudgeStudyTools = (): boolean => {
  const counter = getCounterCookie();
  const numberOfPagesOpenedByUser = getPageCounterCookie();

  return counter < nudgeStudyToolsShowLimit
    && passedTimeInterval()
    && (numberOfPagesOpenedByUser >= nudgeStudyToolsMinPageLimit);
};

// Set required cookies and reset opened page counter
export const setNudgeStudyToolsCookies = () => {
  const now = new Date();
  const counter = getCounterCookie();
  Cookies.set(cookieNudge.counter, (counter + 1).toString(), {expires: daysUntilCookieExpires});
  Cookies.set(cookieNudge.date, now.toString(), {expires: daysUntilCookieExpires});
  Cookies.remove(cookieNudge.pageCounter);
};

export const useIncrementPageCounter = () => {
  const page = useSelector(pageSelector);
  const [counter, updateCounter] = useState<number>(getPageCounterCookie());
  const pageRef = React.useRef<Page>();

  React.useEffect(() => {
    if (page && page !== pageRef.current && counter <= nudgeStudyToolsMinPageLimit) {
      pageRef.current = page;
      updateCounter(incrementPageCounterCookie());
    }
  }, [page, updateCounter, counter]);

  return counter;
};
