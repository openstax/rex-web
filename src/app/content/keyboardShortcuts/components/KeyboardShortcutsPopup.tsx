import { HTMLElement } from '@openstax/types/lib.dom';
import React from 'react';
import { FormattedMessage, useIntl } from 'react-intl';
import { useDispatch, useSelector } from 'react-redux';
import styled from 'styled-components/macro';
import { useAnalyticsEvent } from '../../../../helpers/analytics';
import { useOnEsc, useOnKey } from '../../../reactUtils';
import theme from '../../../theme';
import Modal from '../../components/Modal';
import { bookTheme as bookThemeSelector } from '../../selectors';
import { CloseIcon, CloseIconWrapper, Header } from '../../styles/PopupStyles';
import { closeKeyboardShortcutsMenu, openKeyboardShortcutsMenu } from '../actions';
import * as ksSelectors from '../selectors';
import ShowKeyboardShortcuts from './ShowKeyboardShortcuts';

// tslint:disable-next-line:variable-name
const StyledModal = styled(Modal)`
  max-width: 92.8rem;
`;

// tslint:disable-next-line: variable-name
const KeyboardShortcutsPopup = () => {
  const dispatch = useDispatch();
  const popUpRef = React.useRef<HTMLElement>(null);
  const trackOpenCloseKS = useAnalyticsEvent('openCloseKeyboardShortcuts');
  const isKeyboardShortcutsOpen = useSelector(ksSelectors.isKeyboardShortcutsOpen);
  const bookTheme = useSelector(bookThemeSelector);
  const intl = useIntl();

  const openAndTrack = () => {
    dispatch(openKeyboardShortcutsMenu());
    trackOpenCloseKS();
  };
  useOnKey({key: '?', shiftKey: true}, null, !isKeyboardShortcutsOpen, openAndTrack);

  const closeAndTrack = React.useCallback((method: string) => () => {
    dispatch(closeKeyboardShortcutsMenu());
    trackOpenCloseKS(method);
  }, [trackOpenCloseKS, dispatch]);
  useOnEsc(popUpRef, isKeyboardShortcutsOpen, closeAndTrack('esc'));

  React.useEffect(() => {
    const popUp = popUpRef.current;

    if (popUp && isKeyboardShortcutsOpen) {
      popUp.focus();
    }
  }, [isKeyboardShortcutsOpen]);

  return isKeyboardShortcutsOpen ?
    <StyledModal
      ref={popUpRef}
      tabIndex='-1'
      data-testid='keyboard-shortcuts-popup-wrapper'
      scrollLockProps={{
        mediumScreensOnly: false,
        onClick: closeAndTrack('overlay'),
        overlay: true,
        zIndex: theme.zIndex.keyboardShortcutsPopup,
      }}
    >
      <Header colorSchema={bookTheme}>
        <FormattedMessage id='i18n:a11y:keyboard-shortcuts:heading'>
          {(msg) => msg}
        </FormattedMessage>
        <CloseIconWrapper
          data-testid='close-keyboard-shortcuts-popup'
          aria-label={intl.formatMessage({id: 'i18n:a11y:keyboard-shortcuts:close'})}
          data-analytics-label='Click to close Keyboard Shortcuts modal'
          onClick={closeAndTrack('button')}
        >
          <CloseIcon colorSchema={bookTheme} />
        </CloseIconWrapper>
      </Header>
      <ShowKeyboardShortcuts />
    </StyledModal>
  : null;
};

export default KeyboardShortcutsPopup;
