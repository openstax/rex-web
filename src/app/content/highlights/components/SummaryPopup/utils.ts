import { Highlight } from '@openstax/highlighter/dist/api';
import queryString from 'querystring';
import React from 'react';
import { useSelector } from 'react-redux';
import { systemQueryParameters } from '../../../../navigation/selectors';
import { Book } from '../../../types';
import { findArchiveTreeNodeById } from '../../../utils/archiveTreeUtils';
import { getBookPageUrlAndParams } from '../../../utils/urlUtils';

export const useCreateHighlightLink = ({ sourceId, id, anchor: elementId }: Highlight, book?: Book) => {
  const [link, setLink] = React.useState('');
  const systemQueryParams = useSelector(systemQueryParameters);

  React.useEffect(() => {
    const page = book ? findArchiveTreeNodeById(book.tree, sourceId) : undefined;
    const data: Record<string, string> = {
      ...systemQueryParams,
      target: JSON.stringify({ id, type: 'highlight' }),
    };
    const target = queryString.stringify(data) + `#${elementId}`;
    setLink(`${page && book ? getBookPageUrlAndParams(book, page).url : ''}?${target}`);
  }, [id, book, elementId, sourceId, systemQueryParams]);

  return link;
};
