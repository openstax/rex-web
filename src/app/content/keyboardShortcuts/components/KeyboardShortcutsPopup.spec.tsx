import { HTMLElement } from '@openstax/types/lib.dom';
import React from 'react';
import ReactTestUtils from 'react-dom/test-utils';
import renderer from 'react-test-renderer';
import createTestServices from '../../../../test/createTestServices';
import createTestStore from '../../../../test/createTestStore';
import { dispatchKeyDownEvent, renderToDom } from '../../../../test/reactutils';
import TestContainer from '../../../../test/TestContainer';
import * as navigation from '../../../navigation/selectors';
import { MiddlewareAPI, Store } from '../../../types';
import { assertDocument, assertNotNull } from '../../../utils';
import { content } from '../../routes';
import { closeKeyboardShortcutsMenu, openKeyboardShortcutsMenu } from '../actions';
import * as ksSelectors from '../selectors';
import KeyboardShortcutsPopup from './KeyboardShortcutsPopup';

// hack copied from practiceQuestions popup spec
jest.mock('react', () => {
  const react = (jest as any).requireActual('react');
  return { ...react, useEffect: react.useLayoutEffect };
});

jest.mock('react-dom', () => ({
  ...jest.requireActual('react-dom'),
  createPortal: (children: any) => children,
}));

const mockMatch = {
  params: {
    book: { slug: 'book' },
    page: { slug: 'page' },
  },
  route: content,
  state: {},
};

describe('KeyboardShortcuts', () => {
  let store: Store;
  let services: ReturnType<typeof createTestServices> & MiddlewareAPI;
  let container: HTMLElement;
  let dispatch: jest.SpyInstance;

  beforeEach(() => {
    store = createTestStore();
    services = {
      ...createTestServices(),
      dispatch: store.dispatch,
      getState: store.getState,
    };
    container = assertDocument().createElement('div');
    dispatch = jest.spyOn(store, 'dispatch');
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders keyboard shortcuts modal if it is open', () => {
    jest.spyOn(ksSelectors, 'isKeyboardShortcutsOpen').mockReturnValue(true);

    const component = renderer.create(<TestContainer services={services} store={store}>
      <KeyboardShortcutsPopup />
    </TestContainer>, { createNodeMock: () => container });

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('doesn\'t render keyboard shortcuts modal if it is closed', () => {
    jest.spyOn(ksSelectors, 'isKeyboardShortcutsOpen').mockReturnValue(false);

    const component = renderer.create(<TestContainer services={services} store={store}>
      <KeyboardShortcutsPopup />
    </TestContainer>, { createNodeMock: () => container });

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('focus is on pop up content', async() => {
    jest.spyOn(ksSelectors, 'isKeyboardShortcutsOpen').mockReturnValue(true);
    const focus = jest.fn();
    const addEventListener = jest.fn();
    const createNodeMock = () => ({focus, addEventListener});

    renderer.create(<TestContainer services={services} store={store}>
      <KeyboardShortcutsPopup />
    </TestContainer>, {createNodeMock});

    expect(focus).toHaveBeenCalled();
  });

  it('tracks analytics and adds modal-url when clicking shift + ?', async() => {
    const track = jest.spyOn(services.analytics.openCloseKeyboardShortcuts, 'track');
    jest.spyOn(ksSelectors, 'isKeyboardShortcutsOpen').mockReturnValue(false);
    jest.spyOn(navigation, 'match').mockReturnValue(mockMatch);

    const { node } = renderToDom(<TestContainer services={services} store={store}>
      <KeyboardShortcutsPopup />
    </TestContainer>);
    expect(node).toBeNull();

    expect(track).not.toHaveBeenCalled();

    dispatchKeyDownEvent({key: '?', shiftKey: true});

    expect(track).toHaveBeenCalled();
    expect(dispatch).toHaveBeenCalledWith(openKeyboardShortcutsMenu());
  });

  it('does not interfere with text entry', async() => {
    const track = jest.spyOn(services.analytics.openCloseKeyboardShortcuts, 'track');
    jest.spyOn(ksSelectors, 'isKeyboardShortcutsOpen').mockReturnValue(false);
    jest.spyOn(navigation, 'match').mockReturnValue(mockMatch);

    const { node } = renderToDom(<TestContainer services={services} store={store}>
      <KeyboardShortcutsPopup />
      <input type='search'/>
    </TestContainer>);

    dispatchKeyDownEvent({key: '?', shiftKey: true, target: node});

    expect(track).not.toHaveBeenCalled();
    expect(dispatch).not.toHaveBeenCalled();
  });

  it('tracks analytics and removes modal-url when clicking x icon', () => {
    const track = jest.spyOn(services.analytics.openCloseKeyboardShortcuts, 'track');
    jest.spyOn(ksSelectors, 'isKeyboardShortcutsOpen').mockReturnValue(true);
    jest.spyOn(navigation, 'match').mockReturnValue(mockMatch);

    const component = renderer.create(<TestContainer services={services} store={store}>
      <KeyboardShortcutsPopup />
    </TestContainer>, { createNodeMock: () => container });

    expect(track).not.toHaveBeenCalled();

    renderer.act(() => {
      const closeButton = component.root.findByProps({ 'data-testid': 'close-keyboard-shortcuts-popup' });
      closeButton.props.onClick();
    });

    expect(track).toHaveBeenCalled();
    expect(dispatch).toHaveBeenCalledWith(closeKeyboardShortcutsMenu());
  });

  it('tracks analytics and removes modal-url when clicking esc', async() => {
    const track = jest.spyOn(services.analytics.openCloseKeyboardShortcuts, 'track');
    jest.spyOn(ksSelectors, 'isKeyboardShortcutsOpen').mockReturnValue(true);
    jest.spyOn(navigation, 'match').mockReturnValue(mockMatch);

    const { node } = renderToDom(<TestContainer services={services} store={store}>
      <KeyboardShortcutsPopup />
    </TestContainer>);

    expect(track).not.toHaveBeenCalled();

    const element: HTMLElement = assertNotNull(
      node.querySelector('[data-testid=\'keyboard-shortcuts-popup-wrapper\']'), ''
    );

    dispatchKeyDownEvent({element, key: 'Escape'});

    expect(track).toHaveBeenCalled();
    expect(dispatch).toHaveBeenCalledWith(closeKeyboardShortcutsMenu());
  });

  it('tracks analytics and removes modal-url on overlay click', async() => {
    const track = jest.spyOn(services.analytics.openCloseKeyboardShortcuts, 'track');
    jest.spyOn(ksSelectors, 'isKeyboardShortcutsOpen').mockReturnValue(true);
    jest.spyOn(navigation, 'match').mockReturnValue(mockMatch);

    const { node } = renderToDom(<TestContainer services={services} store={store}>
      <KeyboardShortcutsPopup />
    </TestContainer>);

    expect(track).not.toHaveBeenCalled();

    const element = assertNotNull(node.querySelector('[data-testid=\'scroll-lock-overlay\']'), '');

    const event = assertDocument().createEvent('MouseEvents');
    event.initEvent('click', true, true);
    const preventDefault = event.preventDefault = jest.fn();

    element.dispatchEvent(event); // this checks for bindings using addEventListener
    ReactTestUtils.Simulate.click(element, {preventDefault}); // this checks for react onClick prop

    expect(track).toHaveBeenCalled();
    expect(dispatch).toHaveBeenCalledWith(closeKeyboardShortcutsMenu());
  });
});
