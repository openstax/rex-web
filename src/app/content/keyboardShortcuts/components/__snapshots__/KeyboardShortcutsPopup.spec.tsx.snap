// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KeyboardShortcuts doesn't render keyboard shortcuts modal if it is closed 1`] = `null`;

exports[`KeyboardShortcuts renders keyboard shortcuts modal if it is open 1`] = `
<div
  className="PopupStyles__PopupWrapper-w0z1hl-0 cYcgUJ"
>
  <div
    className="ScrollLock__Overlay-sc-1fi845b-0 gshJxt"
    data-testid="scroll-lock-overlay"
    onClick={[Function]}
  />
  <div
    className="KeyboardShortcutsPopup__StyledModal-y3xath-0 gvXRCW PopupStyles__Modal-w0z1hl-3 dxCHBC"
    data-testid="keyboard-shortcuts-popup-wrapper"
    tabIndex="-1"
  >
    <h3
      className="headings__H3-sc-1xfiggv-2 PopupStyles__Header-w0z1hl-1 kFaPzE"
    >
      Keyboard Shortcuts
      <button
        aria-label="Click to close Keyboard Shortcuts menu"
        className="Button__PlainButton-ayg7nk-2 PopupStyles__CloseIconWrapper-w0z1hl-4 jafVNy"
        data-analytics-label="Click to close Keyboard Shortcuts modal"
        data-testid="close-keyboard-shortcuts-popup"
        onClick={[Function]}
      >
        <svg
          aria-hidden="true"
          className="PopupStyles__CloseIcon-w0z1hl-5 eWfjhO"
          focusable="true"
          height="15px"
          version="1.1"
          viewBox="0 0 15 15"
          width="15px"
        >
          <g
            fill="none"
            fillRule="evenodd"
            stroke="none"
            strokeWidth="1"
          >
            <g
              fill="currentColor"
              transform="translate(-302.000000, -18.000000)"
            >
              <g
                transform="translate(302.000000, 18.000000)"
              >
                <path
                  d="M7.5,5.41522791 L12.0331524,0.579865364 C12.3077536,0.286957429 12.7165503,0.24816296 12.946282,0.493210121 L13.9861449,1.60239723 C14.2158766,1.84744439 14.1795068,2.28349422 13.9049056,2.57640216 L9.37175324,7.41176471 L13.9049056,12.2471273 C14.1795068,12.5400352 14.2158766,12.976085 13.9861449,13.2211322 L12.946282,14.3303193 C12.7165503,14.5753665 12.3077536,14.536572 12.0331524,14.243664 L7.5,9.4083015 L2.96684761,14.243664 C2.69224642,14.536572 2.2834497,14.5753665 2.05371799,14.3303193 L1.01385508,13.2211322 C0.784123363,12.976085 0.820493178,12.5400352 1.09509437,12.2471273 L5.62824676,7.41176471 L1.09509437,2.57640216 C0.820493178,2.28349422 0.784123363,1.84744439 1.01385508,1.60239723 L2.05371799,0.493210121 C2.2834497,0.24816296 2.69224642,0.286957429 2.96684761,0.579865364 L7.5,5.41522791 Z"
                />
              </g>
            </g>
          </g>
        </svg>
      </button>
    </h3>
    <div
      className="PopupStyles__PopupBody-w0z1hl-2 ShowKeyboardShortcuts__ShowKeyboardShortcutsBody-sc-12dalxo-0 eGlfQX"
      data-analytics-region="KS popup"
      data-testid="show-keyboard-shortcuts-body"
    >
      <div
        className="ShowKeyboardShortcuts__ShortcutsHeadingDiv-sc-12dalxo-1 pcTDQ"
      >
        Using a Keyboard for Navigation
      </div>
      <div
        className="ShowKeyboardShortcuts__ShortcutsCard-sc-12dalxo-2 goyzXO"
      >
        <div
          className="ShowKeyboardShortcuts__ShortcutsTable-sc-12dalxo-3 grHFdD"
        >
          <div
            className="ShowKeyboardShortcuts__ShortcutRow-sc-12dalxo-4 iaCHkf"
          >
            <div
              className="ShowKeyboardShortcuts__ShortcutBlock-sc-12dalxo-5 fkJiLz"
            >
              <span
                className="ShowKeyboardShortcuts__ShortcutKey-sc-12dalxo-6 kUZwXn"
              >
                Shift
              </span>
               + 
              <span
                className="ShowKeyboardShortcuts__ShortcutKey-sc-12dalxo-6 kUZwXn"
              >
                ?
              </span>
            </div>
            <div
              className="ShowKeyboardShortcuts__ShortcutBlock-sc-12dalxo-5 fkJiLz"
            >
              Pull up keyboard shortcuts menu
            </div>
          </div>
          <div
            className="ShowKeyboardShortcuts__ShortcutRow-sc-12dalxo-4 iaCHkf"
          >
            <div
              className="ShowKeyboardShortcuts__ShortcutBlock-sc-12dalxo-5 fkJiLz"
            >
              <span
                className="ShowKeyboardShortcuts__ShortcutKey-sc-12dalxo-6 kUZwXn"
              >
                Tab
              </span>
            </div>
            <div
              className="ShowKeyboardShortcuts__ShortcutBlock-sc-12dalxo-5 fkJiLz"
            >
              Move forward through book page elements, highlights and notes
            </div>
          </div>
          <div
            className="ShowKeyboardShortcuts__ShortcutRow-sc-12dalxo-4 iaCHkf"
          >
            <div
              className="ShowKeyboardShortcuts__ShortcutBlock-sc-12dalxo-5 fkJiLz"
            >
              <span
                className="ShowKeyboardShortcuts__ShortcutKey-sc-12dalxo-6 kUZwXn"
              >
                Shift
              </span>
               + 
              <span
                className="ShowKeyboardShortcuts__ShortcutKey-sc-12dalxo-6 kUZwXn"
              >
                Tab
              </span>
            </div>
            <div
              className="ShowKeyboardShortcuts__ShortcutBlock-sc-12dalxo-5 fkJiLz"
            >
              Move backward through book page elements, highlights and notes
            </div>
          </div>
          <div
            className="ShowKeyboardShortcuts__ShortcutRow-sc-12dalxo-4 iaCHkf"
          >
            <div
              className="ShowKeyboardShortcuts__ShortcutBlock-sc-12dalxo-5 fkJiLz"
            >
              <span
                className="ShowKeyboardShortcuts__ShortcutKey-sc-12dalxo-6 kUZwXn"
              >
                h
              </span>
            </div>
            <div
              className="ShowKeyboardShortcuts__ShortcutBlock-sc-12dalxo-5 fkJiLz"
            >
              Move focus into and out of a note
            </div>
          </div>
          <div
            className="ShowKeyboardShortcuts__ShortcutRow-sc-12dalxo-4 iaCHkf"
          >
            <div
              className="ShowKeyboardShortcuts__ShortcutBlock-sc-12dalxo-5 fkJiLz"
            >
              <span
                className="ShowKeyboardShortcuts__ShortcutKey-sc-12dalxo-6 kUZwXn"
              >
                Tab
              </span>
            </div>
            <div
              className="ShowKeyboardShortcuts__ShortcutBlock-sc-12dalxo-5 fkJiLz"
            >
              Move through note editing options
            </div>
          </div>
          <div
            className="ShowKeyboardShortcuts__ShortcutRow-sc-12dalxo-4 iaCHkf"
          >
            <div
              className="ShowKeyboardShortcuts__ShortcutBlock-sc-12dalxo-5 fkJiLz"
            >
              <span
                className="ShowKeyboardShortcuts__ShortcutKey-sc-12dalxo-6 kUZwXn"
              >
                Space
              </span>
            </div>
            <div
              className="ShowKeyboardShortcuts__ShortcutBlock-sc-12dalxo-5 fkJiLz"
            >
              Select a different highlight color
            </div>
          </div>
          <div
            className="ShowKeyboardShortcuts__ShortcutRow-sc-12dalxo-4 iaCHkf"
          >
            <div
              className="ShowKeyboardShortcuts__ShortcutBlock-sc-12dalxo-5 fkJiLz"
            >
              <span
                className="ShowKeyboardShortcuts__ShortcutKey-sc-12dalxo-6 kUZwXn"
              >
                Enter
              </span>
            </div>
            <div
              className="ShowKeyboardShortcuts__ShortcutBlock-sc-12dalxo-5 fkJiLz"
            >
              Save or Cancel note edits
            </div>
          </div>
        </div>
      </div>
      <div
        className="ShowKeyboardShortcuts__ShortcutsHeadingDiv-sc-12dalxo-1 pcTDQ"
      >
        Creating Highlights and Notes with a Keyboard
      </div>
      <div
        className="ShowKeyboardShortcuts__CaretMessageDiv-sc-12dalxo-7 dCQGYV"
        dangerouslySetInnerHTML={
          Object {
            "__html": "This functionality requires the use of a <strong>Caret Browsing extension</strong>, which enables you to browse and select the text of web pages using arrow keys.",
          }
        }
      />
      <div
        className="ShowKeyboardShortcuts__ShortcutsCard-sc-12dalxo-2 goyzXO"
      >
        <div
          className="ShowKeyboardShortcuts__ShortcutsTable-sc-12dalxo-3 grHFdD"
        >
          <div
            className="ShowKeyboardShortcuts__ShortcutRow-sc-12dalxo-4 iaCHkf"
          >
            <div
              className="ShowKeyboardShortcuts__ShortcutBlock-sc-12dalxo-5 fkJiLz"
            >
              <span
                className="ShowKeyboardShortcuts__ShortcutKey-sc-12dalxo-6 kUZwXn"
              >
                Arrows
              </span>
            </div>
            <div
              className="ShowKeyboardShortcuts__ShortcutBlock-sc-12dalxo-5 fkJiLz"
            >
              Move focus through each line of text
            </div>
          </div>
          <div
            className="ShowKeyboardShortcuts__ShortcutRow-sc-12dalxo-4 iaCHkf"
          >
            <div
              className="ShowKeyboardShortcuts__ShortcutBlock-sc-12dalxo-5 fkJiLz"
            >
              <span
                className="ShowKeyboardShortcuts__ShortcutKey-sc-12dalxo-6 kUZwXn"
              >
                Shift
              </span>
               + 
              <span
                className="ShowKeyboardShortcuts__ShortcutKey-sc-12dalxo-6 kUZwXn"
              >
                Arrows
              </span>
            </div>
            <div
              className="ShowKeyboardShortcuts__ShortcutBlock-sc-12dalxo-5 fkJiLz"
            >
              Select text
            </div>
          </div>
          <div
            className="ShowKeyboardShortcuts__ShortcutRow-sc-12dalxo-4 iaCHkf"
          >
            <div
              className="ShowKeyboardShortcuts__ShortcutBlock-sc-12dalxo-5 fkJiLz"
            >
              <span
                className="ShowKeyboardShortcuts__ShortcutKey-sc-12dalxo-6 kUZwXn"
              >
                h
              </span>
            </div>
            <div
              className="ShowKeyboardShortcuts__ShortcutBlock-sc-12dalxo-5 fkJiLz"
            >
              Create a highlight or note after selecting text
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
